From ad612844f37c35be46c90fb465cfb35d01901b63 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Tue, 6 Mar 2018 11:31:14 +0100
Subject: [PATCH 4/7] vold: Make sure block device exists before formatting it

* This fixes race condition where dm-1 doesn't exist
  when doFormat() is being executed on some devices
  while formatting them to use with adopted storage.

Change-Id: I1297025900a988141717b81b9d0fcca7197955bd
---
 Utils.cpp               | 19 +++++++++++++++++++
 Utils.h                 |  3 +++
 model/PrivateVolume.cpp |  7 +++++++
 3 files changed, 29 insertions(+)

diff --git a/Utils.cpp b/Utils.cpp
index 6a3830f..7c87086 100644
--- a/Utils.cpp
+++ b/Utils.cpp
@@ -56,6 +56,8 @@ using namespace std::chrono_literals;
 using android::base::ReadFileToString;
 using android::base::StringPrintf;
 
+using namespace std::chrono_literals;
+
 namespace android {
 namespace vold {
 
@@ -753,6 +755,23 @@ bool Readlinkat(int dirfd, const std::string& path, std::string* result) {
     }
 }
 
+bool WaitForFile(const std::string& filename,
+        const std::chrono::milliseconds relativeTimeout) {
+    auto startTime = std::chrono::steady_clock::now();
+
+    while (true) {
+        if (!access(filename.c_str(), F_OK) || errno != ENOENT) {
+            return true;
+        }
+
+        std::this_thread::sleep_for(50ms);
+
+        auto now = std::chrono::steady_clock::now();
+        auto timeElapsed = std::chrono::duration_cast<std::chrono::milliseconds>(now - startTime);
+        if (timeElapsed > relativeTimeout) return false;
+    }
+}
+
 bool IsRunningInEmulator() {
     return android::base::GetBoolProperty("ro.kernel.qemu", false);
 }
diff --git a/Utils.h b/Utils.h
index c083021..b8f7bd1 100644
--- a/Utils.h
+++ b/Utils.h
@@ -124,6 +124,9 @@ status_t RestoreconRecursive(const std::string& path);
 // TODO: promote to android::base
 bool Readlinkat(int dirfd, const std::string& path, std::string* result);
 
+bool WaitForFile(const std::string& filename,
+        const std::chrono::milliseconds relativeTimeout);
+
 /* Checks if Android is running in QEMU */
 bool IsRunningInEmulator();
 
diff --git a/model/PrivateVolume.cpp b/model/PrivateVolume.cpp
index de2a09f..c9f9725 100644
--- a/model/PrivateVolume.cpp
+++ b/model/PrivateVolume.cpp
@@ -38,6 +38,8 @@
 
 using android::base::StringPrintf;
 
+using namespace std::chrono_literals;
+
 namespace android {
 namespace vold {
 
@@ -188,6 +190,11 @@ status_t PrivateVolume::doFormat(const std::string& fsType) {
         LOG(DEBUG) << "Resolved auto to " << resolvedFsType;
     }
 
+    if (!WaitForFile(mDmDevPath, 15s)) {
+        PLOG(ERROR) << "Timed out waiting for " << getId();
+        return -EIO;
+    }
+
     if (resolvedFsType == "ext4") {
         // TODO: change reported mountpoint once we have better selinux support
         if (ext4::Format(mDmDevPath, 0, "/data")) {
-- 
2.7.4

